### PR TITLE
fix promise support for benchmarks

### PR DESCRIPTION
**Pull request by schnittstabil**
Benchmarks were broken since globby switched to Promises...

### CODE REVIEW

**Comment by SamVerschueren** [comment]
Might be a good idea if we extracte the promise-to-cb code in in separate library?

**Comment by sindresorhus** [comment]
Can't we do the inverse and convert the non-promise APIs to promise?

**Comment by schnittstabil** [comment]
@sindresorhus [matcha](https://github.com/logicalparadox/matcha) doesn't support Promises :(

**Comment by schnittstabil** [comment]
@sindresorhus Is there any reason we need `setImmediate` here?

File: bench.js [hunk 1]
```diff
@@ -8,16 +8,24 @@ var globby = require('./');
...
 var BENCH_DIR = 'bench';
 
+function globbyToCb(globbyFn) {
+	return function (patterns, cb) {
+		setImmediate(function () {
+			globbyFn(patterns).then(function () {
+				cb();
+			}, function (err) {
+				cb(err);
+			});
```
Comment (sindresorhus) [inline bench.js]:
```js
globbyFn(patterns).then(cb.bind(null, null), cb);
```

Comment (schnittstabil) [inline bench.js]: Nice :yum:

**Comment by UltCombo** [comment]
@schnittstabil `setImmediate` was added to workaround a "maximum call stack size exceeded" error in the past, if I recall correctly. I believe that shouldn't happen when you are using promises, so it may be safe to remove.

**Comment by UltCombo** [comment]
More specifically, `setImmediate` was introduced by #12 and the reasoning can be found at https://github.com/isaacs/node-glob/pull/151#issuecomment-69466751.

This was all the way back when globby had a callback API, before the adoption of promises. As you may know, promises are scheduled as [microtasks](https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/) so they are guaranteed to resolve asynchronously. I believe this should never result in a call stack size exceeded error.

So, in short, remove the `setImmediate` and run the tests, if they don't throw a call stack size exceeded error, this is good to go. :smiley:

**Comment by schnittstabil** [comment]
@UltCombo Many thanks; looks like `setImmediate` is no longer needed.

File: bench.js [hunk 2]
```diff
@@ -8,20 +8,16 @@ var globby = require('./');
...
-			globbyFn(patterns).then(cb.bind(null, null), cb);
-		});
-	};
-}
-
 var runners = [{
 	name: 'globby async (working directory)',
-	run: globbyToCb(globby)
+	run: function (patterns, cb) {
+		globby(patterns).then(cb.bind(), cb);
```
Comment (UltCombo) [inline bench.js]: I think you still need `.bind(null, null)` (bind `null` to the callback's `this` and `err` argument), otherwise the promise's fulfillment value will be passed as the error argument to matcha's callback, isn't it?

File: bench.js [hunk 3]
```diff
@@ -8,20 +8,16 @@ var globby = require('./');
...
+	}
 }, {
 	name: 'globby async (upstream/master)',
-	run: globbyToCb(globbyMaster)
+	run: function (patterns, cb) {
+		globbyMaster(patterns).then(cb.bind(), cb);
```
Comment (UltCombo) [inline bench.js]: Same as [above](https://github.com/sindresorhus/globby/pull/23/commits/497c3c7324ee27d58ff36704cc6920c4d233b02a#r56760396).

File: bench.js (hunk 2 above)
Comment (SamVerschueren) [inline bench.js]: :+1:

Comment (schnittstabil) [inline bench.js]: I had first to drink my coffee :sleeping:

**Comment by sindresorhus** [comment]
Awesome. Thanks @schnittstabil :)
